helper/trace/debug_trace.go

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package trace
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime/trace"
+	"time"
+)
+
+func StartDebugTrace(dir string, filePrefix string) (file string, stop func() error, err error) {
+	dirMustExist := true
+	if dir == "" {
+		dirMustExist = false // if a dir is provided it must exist, otherwise we'll create a default one
+		dir = filepath.Join(os.TempDir(), "vault-traces")
+	}
+
+	d, err := os.Stat(dir)
+	if err != nil && !os.IsNotExist(err) {
+		return "", nil, fmt.Errorf("failed to stat trace directory %q: %s", dir, err)
+	}
+
+	if os.IsNotExist(err) && dirMustExist {
+		return "", nil, fmt.Errorf("trace directory %q does not exist", dir)
+	}
+
+	if !os.IsNotExist(err) && !d.IsDir() {
```

Comment:
do we need the `!os.IsNotExist(err)` portion of this condition?

Reply:
yes, I think we need it. If `os.IsNotExist(err)` evaluates to true then we don't want to evaluate `d.IsDir()` since `d` would be `nil`.

The alternative would be to use a nested if-else structure that interleaves error returns with additional operations (creation of default dir), which I don't like as much:
```go
	d, err := os.Stat(dir)
	if err != nil {
		if os.IsNotExist(err) {
			if dirMustExist {
				return "", nil, fmt.Errorf("trace directory %q does not exist", dir)
			} else {
				if err := os.Mkdir(dir, 0o700); err != nil {
					return "", nil, fmt.Errorf("failed to create trace directory %q: %s", dir, err)
				}
			}
		} else {
			return "", nil, fmt.Errorf("failed to stat trace directory %q: %s", dir, err)
		}
	} else if !d.IsDir() {
		return "", nil, fmt.Errorf("trace directory %q is not a directory", dir)
	}
```

Reply:
makes sense, it's good as is!